File: app/src/main/java/com/aadit/wallpapermanager/MainActivity.java
package com.aadit.wallpapermanager;


import android.content.Intent;
import android.net.Uri;
import android.provider.MediaStore;
import android.os.Bundle;

import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import android.app.WallpaperManager;
import android.graphics.Bitmap;
import android.widget.Toast;

import java.io.IOException;


public class MainActivity extends AppCompatActivity {


    public static Button button, button0, button1, button2, button3, button4;
    public static TextView textView4, textView5, textView6;
    private RecyclerView recyclerView;
    private Schedule schedule;
    public ImageView imageView, imageView1;
    private Bitmap current;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);





        textView4 = findViewById(R.id.textView4);
        textView5 = findViewById(R.id.textView5);
        textView6 = findViewById(R.id.textView6);
        imageView = findViewById(R.id.imageView);
        imageView1 = findViewById(R.id.imageView1);
        button = findViewById(R.id.button);
        button0 = findViewById(R.id.button0);
        button1 = findViewById(R.id.button1);
        button2 = findViewById(R.id.button2);
        button3 = findViewById(R.id.button3);
        button4 = findViewById(R.id.button4);
        schedule = new Schedule(MainActivity.this, this);
        current = schedule.getCurrentWallpaper();
        imageView1.setImageBitmap(current);
        button4.setEnabled(false);



        //Select Image
        button0.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                schedule.choosePhotoFromGallery();
            }
        });


        //Start Schedule
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                button.setEnabled(true);
                try {
                    schedule.runner=true;
                    schedule.setWallpaper(textView6, button,button4,button2, button3, button1, imageView1);
                    schedule.thread1.start();

                } catch (Exception e) {
                    Toast.makeText(MainActivity.this, "There has been an error setting the wallpaper", Toast.LENGTH_SHORT).show();
                }

            }
        });
        //Set WP
        button1.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v) {
                try{
                    setSelected();
                }
                catch (IOException e){
                    throw new RuntimeException();
                }
            }
        });

        //Start Time
        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                schedule.buttonName = "start";
                schedule.timePicker(textView4);
                schedule.runner=true;
            }

        });
        //End time
        button3.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                schedule.buttonName = "end";
                schedule.timePicker(textView5);
                schedule.runner=true;
                button4.setEnabled(true);
            }

        });
        //Stop
        button4.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                schedule.runner=false;
                if(schedule.thread1.getState()==Thread.State.RUNNABLE){
                    Toast.makeText(MainActivity.this, "Could not stop process!!", Toast.LENGTH_SHORT).show();
                }
                else{
                    Toast.makeText(MainActivity.this, "Process Stopped.", Toast.LENGTH_SHORT).show();
                }
                schedule.sHour = 0;
                schedule.eHour = 0;
                schedule.sMinute = 0;
                schedule.eMinute = 0;
                textView4.setText("00 : 00");
                textView5.setText("00 : 00");
                button4.setEnabled(false);
            }
        });
    }





    private void setSelected() throws IOException {
        WallpaperManager wallpaperManager = WallpaperManager.getInstance(getApplicationContext());
        wallpaperManager.setBitmap(schedule.selectedImageBitmap);
        textView6.setText("Done ");
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        current = schedule.getCurrentWallpaper();

        if (requestCode == schedule.REQUEST_IMAGE_SELECT && resultCode == RESULT_OK) {
            Uri imageUri = data.getData();

            // Start the image cropping activity
            CropImage.activity(imageUri)
                    .setGuidelines(CropImageView.Guidelines.ON)
                    .setAspectRatio(9,16) // Set the aspect ratio using bitmap dimensions
                    .setCropShape(CropImageView.CropShape.RECTANGLE) // Set the cropping shape to rectangle
                    .setRequestedSize(current.getWidth(), current.getHeight())
                    .setFixAspectRatio(true) // Lock the aspect ratio
                    .start(this);
        } else if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            // Handle the cropped image result
            CropImage.ActivityResult result = CropImage.getActivityResult(data);
            if (resultCode == RESULT_OK) {
                // Get the cropped image URI
                Uri croppedImageUri = result.getUri();

                try {
                    // Load the cropped image into a Bitmap
                    Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContentResolver(), croppedImageUri);

                    // Set the selected image to the ImageView
                    imageView.setImageBitmap(bitmap);

                    // Store the selected image bitmap in the variable
                    schedule.selectedImageBitmap = bitmap;
                } catch (IOException e) {
                    e.printStackTrace();
                }
            } else if (resultCode == CropImage.CROP_IMAGE_ACTIVITY_RESULT_ERROR_CODE) {
                // Handle cropping error
                Exception error = result.getError();
                error.printStackTrace();
            }
        }
    }

}

File: app/src/main/java/com/tutorial/wallpapermanager/MainActivity.kt
package com.tutorial.wallpapermanager

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.canhub.cropper.CropImageContract
import com.canhub.cropper.CropImageContractOptions
import com.canhub.cropper.CropImageOptions
import java.io.InputStream


class MainActivity : AppCompatActivity() {

    lateinit var StartTime: TextView
    lateinit var EndTime: TextView
    lateinit var Status: TextView
    lateinit var selectTextView: TextView
    lateinit var SelectedWP: ImageView
    lateinit var PrevWP: ImageView
    lateinit var buttonStartSchedule: Button
    lateinit var buttonStop: Button

    lateinit var schedule: Schedule
    lateinit var c: Colors
    var selectedImageBitmap: Bitmap? = null



    var selectedImageUri: Uri? = null
    var selectedWallpaperDrawable: Drawable? = null


    val requestPermission: ActivityResultLauncher<Array<String>> =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()){
            permissions ->
            permissions.entries.forEach{
                val permissionName = it.key
                val isGranted = it.value
                if (!isGranted){
                    Toast.makeText(this@MainActivity, "Permission Denied $permissionName", Toast.LENGTH_SHORT).show()
                    showRationaleDialog("Permission Denied", "Permissions are required to use the application. Please grant them manually.")
                    return@registerForActivityResult
                }
            }
        }






    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        requestRequiredPermission()
        StartTime = findViewById(R.id.textView4);
        EndTime = findViewById(R.id.textView5);
        Status = findViewById(R.id.textView6);
        selectTextView = findViewById(R.id.textView2)
        SelectedWP = findViewById(R.id.imageView);
        PrevWP = findViewById(R.id.imageView1);
        buttonStartSchedule = findViewById(R.id.button);
        buttonStop = findViewById(R.id.button4);
        //cropImageView = findViewById(R.id.cropImageView)

        schedule = Schedule(this, this)
        c = Colors(this@MainActivity)
        var borderRed = GradientDrawable()
        borderRed.setColor(Color.TRANSPARENT)
        borderRed.setStroke(2, c.palette2DarkRed)



        buttonStartSchedule.isEnabled = false
        Schedule.buttonStart = buttonStartSchedule
        SelectedWP.background = borderRed
        selectTextView.setTextColor(c.palette2DarkRed)

        StartTime.setTextColor(c.palette2DarkRed)
        EndTime.setTextColor(c.palette2DarkRed)

        requestRequiredPermission()
        consistencyCheck()




        //https://github.com/CanHub/Android-Image-Cropper



        val cropImage = registerForActivityResult(CropImageContract()) { result ->
            if (result.isSuccessful) {
                // Use the returned uri.
                val uriContent = result.uriContent
                selectedImageBitmap = resizeBitmap(uriToBitmap(uriContent, this))
                Schedule.selectedWallpaperBitmap = selectedImageBitmap
                SelectedWP.setImageBitmap(Schedule.selectedWallpaperBitmap)
                selectTextView.setText("Wallpaper Selected")
                selectTextView.setTextColor(c.palette4green)

            } else {
                // An error occurred.
                val exception = result.error
            }
        }

        val pickImage = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                val data: Intent? = result.data
                if (data != null) {
                    val uri: Uri? = data.data
                    if (uri != null) {
                        selectedImageUri = uri
                        var cio = CropImageOptions()
                        cio.allowRotation = true
                        cio.fixAspectRatio = true
                        cio.aspectRatioY = 16
                        cio.aspectRatioX = 9
                        cio.activityTitle = "Crop Image"

                        cropImage.launch(
                            CropImageContractOptions(selectedImageUri, cio)
                        )


                    }
                }
            }
        }


        SelectedWP.setOnClickListener{
            val galleryIntent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
            pickImage.launch(galleryIntent)
        }


        StartTime.setOnClickListener{
            schedule.showTimePicker("s", StartTime)
        }
        EndTime.setOnClickListener{
            schedule.showTimePicker("e", EndTime)
        }
        buttonStartSchedule.setOnClickListener {
            if(!errorFound()){
                val intent = Intent(applicationContext, CountDownService::class.java)
                Schedule.intentPassed = intent
                Toast.makeText(this, "${Schedule.startHour} : ${Schedule.startMin} to ${Schedule.endHour} : ${Schedule.endMin}", Toast.LENGTH_SHORT).show()
                intent.putExtra("startTime", arrayOf(Schedule.startHour, Schedule.startMin))
                intent.putExtra("signal", "start")
                intent.action = CountDownService.Actions.START.toString()
                startService(intent)
                schedule.doWithIntent(intent)
            }


        }
        
        buttonStop.setOnClickListener { 
            if(Schedule.intentPassed == null){
                Toast.makeText(this, "The schedule has not started yet or has been completed", Toast.LENGTH_SHORT).show()
            }
            else{
//                Schedule.intentPassed!!.action = CountDownService.Actions.STOP.toString()
//                stopService(Schedule.intentPassed)
//                Toast.makeText(this, "Scheduled job has been cancelled.", Toast.LENGTH_SHORT).show()
                Schedule.endService(Schedule.intentPassed!!)
            }
        }

    }

    private fun consistencyCheck(){
        if(Schedule.wallpaperChangeFlag==false){
            schedule.getCurrentWallpaper()
            Schedule.wallpaperChangeFlag=true

        }
        if(Schedule.selectedWallpaperBitmap != null){
            SelectedWP.setImageBitmap(Schedule.selectedWallpaperBitmap)
        }
        if(Schedule.startTime != null){
            StartTime.text = Schedule.startTime
            StartTime.setTextColor(c.palette4green)
        }
        if(Schedule.endTime != null){
            EndTime.text = Schedule.endTime
            EndTime.setTextColor(c.palette4green)
            buttonStartSchedule.isEnabled = true
        }
        PrevWP.setImageBitmap(Schedule.prevWallpaperBitmap)
    }

    private fun errorFound(): Boolean{
        if(Schedule.selectedWallpaperBitmap == null){
            Toast.makeText(this, "Select a wallpaper to set", Toast.LENGTH_SHORT).show()
            return true
        }
        if(Schedule.startHour.toInt() == 0 && Schedule.startMin.toInt() == 0){
            Toast.makeText(this, "Select a time for the schedule to start", Toast.LENGTH_SHORT).show()
            return true
        }
        if(Schedule.endMin.toInt() == 0 && Schedule.endHour.toInt() == 0){
            Toast.makeText(this, "Select a time for the schedule to end", Toast.LENGTH_SHORT).show()
            return true
        }
        return false
    }


    private fun requestRequiredPermission(){
            requestPermission.launch(
                arrayOf(
                    //Manifest.permission.MANAGE_EXTERNAL_STORAGE,
                    Manifest.permission.FOREGROUND_SERVICE,
                    Manifest.permission.READ_EXTERNAL_STORAGE,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE,
                    Manifest.permission.SET_WALLPAPER
                )
            )
    }

    private fun showRationaleDialog(
        title: String,
        message: String
    ) {
        val builder: AlertDialog.Builder = AlertDialog.Builder(this)
        builder.setTitle(title)
            .setMessage(message)
            .setNegativeButton("Cancel") { dialog, _ ->
                dialog.dismiss()
                finish()
            }
        builder.create().show()
    }

    private fun uriToDrawable(uri: Uri, context: Context): Drawable?{
        try{
            val inputStream: InputStream? = context.contentResolver.openInputStream(uri)
            return Drawable.createFromStream(inputStream, uri.toString())
        }
        catch(e: Exception){
            e.printStackTrace()
        }
        return null
    }

    private fun uriToBitmap(uri: Uri?, context: Context): Bitmap {

        val inputStream: InputStream? = uri?.let { context.contentResolver.openInputStream(it) }
        return BitmapFactory.decodeStream(inputStream)

    }


    private fun resizeBitmap(bitmap: Bitmap): Bitmap {
        val originalWidth = bitmap.width
        val originalHeight = bitmap.height
        val maxWidth = 1080
        val aspectRatio: Double = 9.0/16.0
        // Calculate the new height based on the maximum width and aspect ratio
        val newWidth = if (originalWidth > maxWidth) maxWidth else originalWidth
        val newHeight = (newWidth / aspectRatio).toInt()

        // Resize the bitmap
        return Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
    }

    private fun checkRunning(){

    }




}